### ISSUE TITLE
New rule: Require regex character sets ranges to be obvious (A-Z, not A-z)

### ISSUE THREAD

**Issue by xfix** [eslint/eslint#8740]
**Please describe what the rule should do:**

When seeing a range in a regular expression (for example: `/[a-z]/`) it should require they are subsets of one of the following ranges:

- `a-z`
- `A-Z`
- `0-9`

This prevents mistakes like `A-z` that will work, but match too much. For instance, `/[A-z]/.test("^")` returns `true`, which may be unexpected. Additionally, it warns about unintended uses of range operator, for instance when using a regular expression to parse grammar (`-` needs to be specifically escaped).

This rule shouldn't work on unicode/hex escapes, as using those shows that the string may actually store binary data. It also shouldn't work on characters outside of ASCII range, as there are legitimate use of those (for instance, matching codepoint blocks like Arabic), and it would require thinking about what is reasonable as far this rule is concerned - it can be extended later if needed. Also, as I mentioned, it requires a subset, so code like `/[d-g]/` is perfectly fine.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
const OPERATOR_REGEX = /[+-\/*]/
```

```js
const word = "Hello^world"
if (word.test(/^[A-z]+$/)) {
    alert("An identifier")
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

Writing regular expression is hard. Ranges like `[A-z]` are quietly wrong, even if they appear to work. It's likely to forget to escape `-` in ranges, and get confused why the code doesn't work (happened a lot back when I was learning regular expressions, I do admit).

This shouldn't be a plugin, as likely a plugin to check this wouldn't be used as it wouldn't be known by users of ESLint, while this is a rather serious error which may be confusing to debug. This rule is very unlikely to be a false positive, and in event it turns out to be, it's easy to use hex escapes. Regular expressions are part of a language, not an external library.

There is [prior art for this lint in Perl](http://perldoc.perl.org/perldiag.html#Ranges-of-ASCII-printables-should-be-some-subset-of-%220-9%22%2c-%22A-Z%22%2c-or-%22a-z%22-in-regex%3b-marked-by-%3c---HERE-in-m%2f%25s%2f).

I did to propose this rule after [realizing that `no-useless-escape` complains about `\-` in character ranges, when `-` is at beginning or end of regex](https://github.com/Zarel/Pokemon-Showdown/pull/3637#issuecomment-308811837). I think `no-useless-escape` lint rule is fine, as long it is supported by another rule finding mistakes with regex ranges, as encouraging to not write `\-` makes such mistakes more likely

**Comment by not-an-aardvark** [comment]
Thanks for the proposal.

This seems like it could be useful -- I've accidentally created a [bug](https://github.com/eslint/eslint/pull/7105#pullrequestreview-2469827) as a result of this before.

However, would it allow unicode ranges? For example, one could use this regex to test for ASCII characters, and I would say it is not confusing in the same way as the other examples:

```js
/[\u0000-\u001f]/
```

**Comment by Zarel** [comment]
Yes, allowing ranges does appear to be the implication of "in event it turns out to be, it's easy to use hex escapes"

**Comment by xfix** [comment]
I wrote "This rule shouldn't work on unicode/hex escapes, as using those shows that the string may actually store binary data.", but the exceptions to the rule probably should have been more noticeable in my issue. Oh well.

**Comment by not-an-aardvark** [comment]
Sorry, not sure how I missed that. In that case this sounds good to me.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

**Comment by Zarel** [comment]
It seems like this issue did reach consensus? Was your closing a mistake?
